pkg/common-controller/snapshot_controller.go

```diff
@@ -1591,3 +1602,31 @@ func (ctrl *csiSnapshotCommonController) checkAndSetInvalidSnapshotLabel(snapsho
 
 	return updatedSnapshot, nil
 }
+
+func (ctrl *csiSnapshotCommonController) getManagedByNode(snapshot *crdv1.VolumeSnapshot) (string, error) {
+	pv, err := ctrl.getVolumeFromVolumeSnapshot(snapshot)
+	if err != nil {
+		return "", fmt.Errorf("failed to retrieve PV from snapshot[%s]: %q", snapshot.Name, err)
+	}
+	if pv.Spec.NodeAffinity == nil {
+		klog.V(5).Infof("NodeAffinity not set for pv %s", pv.Name)
+		return "", nil
+	}
+	nodeSelectorTerms := pv.Spec.NodeAffinity.Required
+
+	nodes, err := ctrl.client.CoreV1().Nodes().List(context.Background(), metav1.ListOptions{})
+	if err != nil {
+		klog.Errorf("failed to get the list of nodes: %q", err)
+		return "", nil
+	}
+
+	for _, node := range nodes.Items {
```

Comment:
The fact that the first matching node will ALWAYS be picked to be labeled is a bit concerning though TBH I do not have a solution. It might result in a situation all snapshot contents will be labeled with the same node name, and the sidecar running on that node became hot.

Reply:
Can there be multiple nodes that can match with the node affinity for a local volume?

Reply:
AFAIK there can be only a single node that matches the node affinity